File: app/src/main/java/com/fstyle/structure_android/screen/userdetail/UserDetailPresenter.java
package com.fstyle.structure_android.screen.userdetail;

import com.fstyle.structure_android.data.repository.UserRepository;
import com.fstyle.structure_android.utils.rx.BaseSchedulerProvider;
import com.google.common.base.Preconditions;

/**
 * Listens to user actions from the UI ({@link UserDetailActivity}), retrieves the data and updates
 * the UI as required.
 */
final class UserDetailPresenter implements UserDetailContract.Presenter {
    private static final String TAG = UserDetailPresenter.class.getName();

    private UserDetailContract.View mView;

    private UserRepository mUserRepository;
    private BaseSchedulerProvider mSchedulerProvider;

    UserDetailPresenter(UserRepository userRepository,
            BaseSchedulerProvider schedulerProvider) {
        mUserRepository = Preconditions.checkNotNull(userRepository);
        mSchedulerProvider = Preconditions.checkNotNull(schedulerProvider);
    }

    @Override
    public void setView(UserDetailContract.View view) {
        mView = view;
    }

    @Override
    public void onStart() {
        loadUser();
    }

    private void loadUser() {
        mUserRepository.getUser(mView.getAvatarUrl())
                .subscribeOn(mSchedulerProvider.io())
                .observeOn(mSchedulerProvider.ui())
                .subscribe(user -> mView.showUser(user), error -> mView.showGetUserError(error));
    }

    @Override
    public void onStop() {
    }
}

File: app/src/main/java/com/fstyle/structure_android/screen/userdetail/UserDetailPresenter.kt
package com.fstyle.structure_android.screen.userdetail;

import com.fstyle.structure_android.data.source.UserRepository
import com.fstyle.structure_android.data.source.remote.api.error.BaseException
import com.fstyle.structure_android.utils.rx.BaseSchedulerProvider

/**
 * Listens to user actions from the UI ({@link UserDetailActivity}), retrieves the data and updates
 * the UI as required.
 */
class UserDetailPresenter(
    private val userRepository: UserRepository) : UserDetailContract.Presenter {

  private var schedulerProvider: BaseSchedulerProvider? = null
  private var mViewModel: UserDetailContract.ViewModel? = null

  override fun onStart() {}

  override fun onStop() {}

  override fun setViewModel(viewModel: UserDetailContract.ViewModel) {
    mViewModel = viewModel
  }

  override fun setSchedulerProvider(schedulerProvider: BaseSchedulerProvider?) {
    this.schedulerProvider = schedulerProvider
  }

  override fun getUserDetailFromServer(userLogin: String?) {
    userRepository.getUserDetailFromServer(userLogin)
        .subscribeOn(schedulerProvider?.io())
        .doOnSubscribe { mViewModel?.onShowProgressBar() }
        .observeOn(schedulerProvider?.ui())
        .doAfterTerminate { mViewModel?.onHideProgressBar() }
        .subscribe({ user -> mViewModel?.onGetUserDetailSuccess(user) },
            { error -> mViewModel?.onGetUserDetailError(error as BaseException) })
  }
}
